### PR TITLE
Allow Nyxt to load extensions

### PR DESCRIPTION
**Pull request by jmercouris**
### Synopsis
I've added the following to the end of `start.lisp`:

```
(map () 'asdf:register-preloaded-system (asdf:already-loaded-systems))
(map () 'asdf:register-immutable-system (asdf:already-loaded-systems))
```

This prevents attempted recompilation of already loaded systems in an image of Nyxt. This is critical for loading on NixOS. It prevents ASDF from attempting to recompile systems that it cannot (they are located in non-writable directories).

This therefore allows you to do:

```
(asdf:load-asd "/home/jmercouris/Source/Lisp/nx-kaomoji/nx-kaomoji.asd")
(asdf:load-system :nx-kaomoji :force-not t)
```

notice the key argument of `:force-not`. This is important. You can read more about it in the ASDF source code. Here is the documentation from `asdf::*immutable-systems*`:

```
"A hash-set (equal hash-table mapping keys to T) of systems that are immutable,
i.e. already loaded in memory and not to be refreshed from the filesystem.
They will be treated specially by find-system, and passed as :force-not argument to make-plan.

For instance, to can deliver an image with many systems precompiled, that *will not* check the
filesystem for them every time a user loads an extension, what more risk a problematic upgrade
 or catastrophic downgrade, before you dump an image, you may use:
   (map () 'asdf:register-immutable-system (asdf:already-loaded-systems))

Note that direct access to this variable from outside ASDF is not supported.
Please call REGISTER-IMMUTABLE-SYSTEM to add new immutable systems, and
contact maintainers if you need a stable API to do more than that."
```

### Questions
Please let me know if this also works for Guix. I believe it should also alleviate any problems on Ubuntu as well, though I do not have a deb to test against.

Also, please let me know if you can think of a better place to add (perhaps it belongs somewhere in the asd?):

```
(map () 'asdf:register-preloaded-system (asdf:already-loaded-systems))
(map () 'asdf:register-immutable-system (asdf:already-loaded-systems))
```

### CODE REVIEW

**Comment by jmercouris** [comment]
If you complete this quest, the nectar that awaits you at the end of this journey is unfettered access to `nx-kaomoji`. I can think of no sweeter prize.

Screenshot_20210505_172755

**Comment by Ambrevar** [comment]
Indeed the code belongs to the dump hook I think, as mentioned in the documentation.

For instance we could set `uiop:*image-dump-hook*` to this in the `nyxt.asd` file.
Needs some testing.

This could be a solution for Nix.
The problem with the Guix package is different if I recall what @aartaka reported.  I'll investigate on that end.

File: source/start.lisp [hunk 1]
```diff
@@ -532,3 +532,6 @@ Finally,run the `*after-init-hook*'."
 (define-command nyxt-init-time ()
   "Return the duration of Nyxt initialization."
   (echo "~,2f seconds" (slot-value *browser* 'init-time)))
+
+(map () 'asdf:register-preloaded-system (asdf:already-loaded-systems))
+(map () 'asdf:register-immutable-system (asdf:already-loaded-systems))
```
[the remark this answers is not in the document]
Comment (jmercouris) [reply source/start.lisp:537]: I'm just using the example provided by ASDF, not sure why they wrote it that way.

**Comment by jmercouris** [comment]
In any case, I think this is important. Compiled images of Nyxt should not be recompiling already loaded dependencies!

**Comment by Ambrevar** [comment]
Instead of the hook, we could also add a `:before` method to `image-op`'s `perform`.  See what we have already for an example:

```lisp
#+sb-core-compression
(handler-bind ((warning #'muffle-warning))
  (defmethod perform ((o image-op) (c system))
    (uiop:dump-image (output-file o c)
                     :executable t
                     :compression (uiop:getenv "NYXT_COMPRESS"))))
```

**Comment by jmercouris** [comment]
OK, I'll try that.

**Comment by jmercouris** [comment]
Well it seems that works: (-‿◦☀)

**Comment by jmercouris** [comment]
we are now ready to rock: `ʕ•̫͡•ʔ♬✧`

File: nyxt.asd [hunk 2]
```diff
@@ -311,6 +311,12 @@
                      :executable t
                      :compression (uiop:getenv "NYXT_COMPRESS"))))
 
+(defmethod perform :before ((o image-op) (c system))
+  ;; Register preloaded/immutable systems to prevent compiled images
+  ;; of Nyxt from trying to recompile dependencies
```
Comment (Ambrevar) [inline nyxt.asd:316]: Cool!
Can you turn the comment into a docstring, and add a reference to the ASDF docstring / documentation where you found the code?

**Comment by Ambrevar** [comment]
I'll test on Guix.
I'll also look at how StumpWM does it.

**Comment by Ambrevar** [comment]
For past references, see #1080.

**Comment by Ambrevar** [comment]
In the meantime, can you trigger the deb build on this branch and test it?

**Comment by Ambrevar** [comment]
Nit: Please squash the commits :)

**Comment by Ambrevar** [comment]
Ma, look what I found! :)

https://github.com/stumpwm/stumpwm/blob/master/make-image.lisp.in

**Comment by Ambrevar** [comment]
Why call asdf:register-preloaded-system?
It's not in the doc, nor used by StumpWM.  I suggest to remove it.

**Comment by Ambrevar** [comment]
It works in Guix!

If I'm not mistaken, all that remains to be test is with Ubuntu.
The results could be slightly different because of the use of
linux-packaging.
Fingers crossed.

Should I generate the .deb?

**Comment by jmercouris** [comment]
Please generate the .deb, I'm not sure how to trigger the artifact generation.

File: nyxt.asd (hunk 2 above)
Comment (jmercouris) [reply nyxt.asd:316]: I can't do that, ASDF is a single file. I can make it a docstring and squash the commits.

**Comment by jmercouris** [comment]
> Why call asdf:register-preloaded-system? It's not in the doc, nor used by StumpWM. I suggest to remove it.

I was advised to do it, and it makes things work on NixOS for some reason :-)

**Comment by Ambrevar** [comment]
>> Why call asdf:register-preloaded-system? It's not in the doc, nor used by StumpWM. I suggest to remove it.
>
> I was advised to do it, and it makes things work on NixOS for some reason :-)

Can you confirm it does not work without it?
Then please add a comment explaining that it's required for NixOS and
that it needs more investigation.

**Comment by jmercouris** [comment]
I tried without it, and it did not work, but I will try again :-D

File: nyxt.asd (hunk 2 above)
Comment (Ambrevar) [reply nyxt.asd]: Can't do what?  Refer to the asdf docstring?  Of course, just write `asdf::*immutable-systems*` in the perform docstring.

**Comment by Ambrevar** [comment]
Thanks!

**Comment by Ambrevar** [comment]
I'll trigger the .deb build, don't push anything.

**Comment by jmercouris** [comment]
No idea what changed, perhaps there was some state in emacs, but you are correct, that is indeed not necessary...!

**Comment by jmercouris** [comment]
OK, I won't push anything.

**Comment by Ambrevar** [comment]
See https://github.com/atlas-engineer/nyxt/pull/1366/commits/bdaef5f38a000b700f293cdad3b87cb7951e3e1b for how to trigger the .deb build.

**Comment by jmercouris** [comment]
`(>^_^)><(^o^<)`

**Comment by jmercouris** [comment]
I see, thanks!

**Comment by jmercouris** [comment]
Works on Ubuntu!

Screenshot_20210505_212534

**Comment by jmercouris** [comment]
Merged with 5386223c

**Comment by Ambrevar** [comment]
Fantastic!  And thanks again for taking the pain to deal with the VM!

**Comment by jmercouris** [comment]
No problem!

**Comment by aartaka** [comment]
It work! Thanks John! `＼（＠￣∇￣＠）／`

**Comment by danrobi11** [comment]
@jmercouris @Ambrevar  ❤️

**Comment by jmercouris** [comment]
Thank you :-)

**Pull request closed without merging**